// moment/src/lib/units/day-of-week.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const format_1 = require("../format/format");
const aliases_1 = require("./aliases");
const priorities_1 = require("./priorities");
const regex_1 = require("../parse/regex");
const token_1 = require("../parse/token");
const to_int_1 = require("../utils/to-int");
const is_array_1 = require("../utils/is-array");
const index_of_1 = require("../utils/index-of");
const has_own_prop_1 = require("../utils/has-own-prop");
const utc_1 = require("../create/utc");
const parsing_flags_1 = require("../create/parsing-flags");
// FORMATTING
format_1.addFormatToken('d', 0, 'do', 'day');
format_1.addFormatToken('dd', 0, 0, function (format) {
    return this.localeData().weekdaysMin(this, format);
});
format_1.addFormatToken('ddd', 0, 0, function (format) {
    return this.localeData().weekdaysShort(this, format);
});
format_1.addFormatToken('dddd', 0, 0, function (format) {
    return this.localeData().weekdays(this, format);
});
format_1.addFormatToken('e', 0, 0, 'weekday');
format_1.addFormatToken('E', 0, 0, 'isoWeekday');
// ALIASES
aliases_1.addUnitAlias('day', 'd');
aliases_1.addUnitAlias('weekday', 'e');
aliases_1.addUnitAlias('isoWeekday', 'E');
// PRIORITY
priorities_1.addUnitPriority('day', 11);
priorities_1.addUnitPriority('weekday', 11);
priorities_1.addUnitPriority('isoWeekday', 11);
// PARSING
regex_1.addRegexToken('d', regex_1.match1to2);
regex_1.addRegexToken('e', regex_1.match1to2);
regex_1.addRegexToken('E', regex_1.match1to2);
regex_1.addRegexToken('dd', function (isStrict, locale) {
    return locale.weekdaysMinRegex(isStrict);
});
regex_1.addRegexToken('ddd', function (isStrict, locale) {
    return locale.weekdaysShortRegex(isStrict);
});
regex_1.addRegexToken('dddd', function (isStrict, locale) {
    return locale.weekdaysRegex(isStrict);
});
token_1.addWeekParseToken(['dd', 'ddd', 'dddd'], function (input, week, config, token) {
    var weekday = config._locale.weekdaysParse(input, token, config._strict);
    // if we didn't get a weekday name, mark the date as invalid
    if (weekday != null) {
        week.d = weekday;
    }
    else {
        parsing_flags_1.default(config).invalidWeekday = input;
    }
});
token_1.addWeekParseToken(['d', 'e', 'E'], function (input, week, config, token) {
    week[token] = to_int_1.default(input);
});
// HELPERS
function parseWeekday(input, locale) {
    if (typeof input !== 'string') {
        return input;
    }
    if (!isNaN(input)) {
        return parseInt(input, 10);
    }
    input = locale.weekdaysParse(input);
    if (typeof input === 'number') {
        return input;
    }
    return null;
}
function parseIsoWeekday(input, locale) {
    if (typeof input === 'string') {
        return locale.weekdaysParse(input) % 7 || 7;
    }
    return isNaN(input) ? null : input;
}
// LOCALES
exports.defaultLocaleWeekdays = 'Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday'.split('_');
function localeWeekdays(m, format) {
    if (!m) {
        return is_array_1.default(this._weekdays) ? this._weekdays :
            this._weekdays['standalone'];
    }
    return is_array_1.default(this._weekdays) ? this._weekdays[m.day()] :
        this._weekdays[this._weekdays.isFormat.test(format) ? 'format' : 'standalone'][m.day()];
}
exports.localeWeekdays = localeWeekdays;
exports.defaultLocaleWeekdaysShort = 'Sun_Mon_Tue_Wed_Thu_Fri_Sat'.split('_');
function localeWeekdaysShort(m) {
    return (m) ? this._weekdaysShort[m.day()] : this._weekdaysShort;
}
exports.localeWeekdaysShort = localeWeekdaysShort;
exports.defaultLocaleWeekdaysMin = 'Su_Mo_Tu_We_Th_Fr_Sa'.split('_');
function localeWeekdaysMin(m) {
    return (m) ? this._weekdaysMin[m.day()] : this._weekdaysMin;
}
exports.localeWeekdaysMin = localeWeekdaysMin;
function handleStrictParse(weekdayName, format, strict) {
    var i, ii, mom, llc = weekdayName.toLocaleLowerCase();
    if (!this._weekdaysParse) {
        this._weekdaysParse = [];
        this._shortWeekdaysParse = [];
        this._minWeekdaysParse = [];
        for (i = 0; i < 7; ++i) {
            mom = utc_1.createUTC([2000, 1]).day(i);
            this._minWeekdaysParse[i] = this.weekdaysMin(mom, '').toLocaleLowerCase();
            this._shortWeekdaysParse[i] = this.weekdaysShort(mom, '').toLocaleLowerCase();
            this._weekdaysParse[i] = this.weekdays(mom, '').toLocaleLowerCase();
        }
    }
    if (strict) {
        if (format === 'dddd') {
            ii = index_of_1.default.call(this._weekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
        else if (format === 'ddd') {
            ii = index_of_1.default.call(this._shortWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
        else {
            ii = index_of_1.default.call(this._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
    }
    else {
        if (format === 'dddd') {
            ii = index_of_1.default.call(this._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._shortWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
        else if (format === 'ddd') {
            ii = index_of_1.default.call(this._shortWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
        else {
            ii = index_of_1.default.call(this._minWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = index_of_1.default.call(this._shortWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
    }
}
function localeWeekdaysParse(weekdayName, format, strict) {
    var i, mom, regex;
    if (this._weekdaysParseExact) {
        return handleStrictParse.call(this, weekdayName, format, strict);
    }
    if (!this._weekdaysParse) {
        this._weekdaysParse = [];
        this._minWeekdaysParse = [];
        this._shortWeekdaysParse = [];
        this._fullWeekdaysParse = [];
    }
    for (i = 0; i < 7; i++) {
        // make the regex if we don't have it already
        mom = utc_1.createUTC([2000, 1]).day(i);
        if (strict && !this._fullWeekdaysParse[i]) {
            this._fullWeekdaysParse[i] = new RegExp('^' + this.weekdays(mom, '').replace('.', '\.?') + '$', 'i');
            this._shortWeekdaysParse[i] = new RegExp('^' + this.weekdaysShort(mom, '').replace('.', '\.?') + '$', 'i');
            this._minWeekdaysParse[i] = new RegExp('^' + this.weekdaysMin(mom, '').replace('.', '\.?') + '$', 'i');
        }
        if (!this._weekdaysParse[i]) {
            regex = '^' + this.weekdays(mom, '') + '|^' + this.weekdaysShort(mom, '') + '|^' + this.weekdaysMin(mom, '');
            this._weekdaysParse[i] = new RegExp(regex.replace('.', ''), 'i');
        }
        // test the regex
        if (strict && format === 'dddd' && this._fullWeekdaysParse[i].test(weekdayName)) {
            return i;
        }
        else if (strict && format === 'ddd' && this._shortWeekdaysParse[i].test(weekdayName)) {
            return i;
        }
        else if (strict && format === 'dd' && this._minWeekdaysParse[i].test(weekdayName)) {
            return i;
        }
        else if (!strict && this._weekdaysParse[i].test(weekdayName)) {
            return i;
        }
    }
}
exports.localeWeekdaysParse = localeWeekdaysParse;
// MOMENTS
function getSetDayOfWeek(input) {
    if (!this.isValid()) {
        return input != null ? this : NaN;
    }
    var day = this._isUTC ? this._d.getUTCDay() : this._d.getDay();
    if (input != null) {
        input = parseWeekday(input, this.localeData());
        return this.add(input - day, 'd');
    }
    else {
        return day;
    }
}
exports.getSetDayOfWeek = getSetDayOfWeek;
function getSetLocaleDayOfWeek(input) {
    if (!this.isValid()) {
        return input != null ? this : NaN;
    }
    var weekday = (this.day() + 7 - this.localeData()._week.dow) % 7;
    return input == null ? weekday : this.add(input - weekday, 'd');
}
exports.getSetLocaleDayOfWeek = getSetLocaleDayOfWeek;
function getSetISODayOfWeek(input) {
    if (!this.isValid()) {
        return input != null ? this : NaN;
    }
    // behaves the same as moment#day except
    // as a getter, returns 7 instead of 0 (1-7 range instead of 0-6)
    // as a setter, sunday should belong to the previous week.
    if (input != null) {
        var weekday = parseIsoWeekday(input, this.localeData());
        return this.day(this.day() % 7 ? weekday : weekday - 7);
    }
    else {
        return this.day() || 7;
    }
}
exports.getSetISODayOfWeek = getSetISODayOfWeek;
var defaultWeekdaysRegex = regex_1.matchWord;
function weekdaysRegex(isStrict) {
    if (this._weekdaysParseExact) {
        if (!has_own_prop_1.default(this, '_weekdaysRegex')) {
            computeWeekdaysParse.call(this);
        }
        if (isStrict) {
            return this._weekdaysStrictRegex;
        }
        else {
            return this._weekdaysRegex;
        }
    }
    else {
        if (!has_own_prop_1.default(this, '_weekdaysRegex')) {
            this._weekdaysRegex = defaultWeekdaysRegex;
        }
        return this._weekdaysStrictRegex && isStrict ?
            this._weekdaysStrictRegex : this._weekdaysRegex;
    }
}
exports.weekdaysRegex = weekdaysRegex;
var defaultWeekdaysShortRegex = regex_1.matchWord;
function weekdaysShortRegex(isStrict) {
    if (this._weekdaysParseExact) {
        if (!has_own_prop_1.default(this, '_weekdaysRegex')) {
            computeWeekdaysParse.call(this);
        }
        if (isStrict) {
            return this._weekdaysShortStrictRegex;
        }
        else {
            return this._weekdaysShortRegex;
        }
    }
    else {
        if (!has_own_prop_1.default(this, '_weekdaysShortRegex')) {
            this._weekdaysShortRegex = defaultWeekdaysShortRegex;
        }
        return this._weekdaysShortStrictRegex && isStrict ?
            this._weekdaysShortStrictRegex : this._weekdaysShortRegex;
    }
}
exports.weekdaysShortRegex = weekdaysShortRegex;
var defaultWeekdaysMinRegex = regex_1.matchWord;
function weekdaysMinRegex(isStrict) {
    if (this._weekdaysParseExact) {
        if (!has_own_prop_1.default(this, '_weekdaysRegex')) {
            computeWeekdaysParse.call(this);
        }
        if (isStrict) {
            return this._weekdaysMinStrictRegex;
        }
        else {
            return this._weekdaysMinRegex;
        }
    }
    else {
        if (!has_own_prop_1.default(this, '_weekdaysMinRegex')) {
            this._weekdaysMinRegex = defaultWeekdaysMinRegex;
        }
        return this._weekdaysMinStrictRegex && isStrict ?
            this._weekdaysMinStrictRegex : this._weekdaysMinRegex;
    }
}
exports.weekdaysMinRegex = weekdaysMinRegex;
function computeWeekdaysParse() {
    function cmpLenRev(a, b) {
        return b.length - a.length;
    }
    var minPieces = [], shortPieces = [], longPieces = [], mixedPieces = [], i, mom, minp, shortp, longp;
    for (i = 0; i < 7; i++) {
        // make the regex if we don't have it already
        mom = utc_1.createUTC([2000, 1]).day(i);
        minp = this.weekdaysMin(mom, '');
        shortp = this.weekdaysShort(mom, '');
        longp = this.weekdays(mom, '');
        minPieces.push(minp);
        shortPieces.push(shortp);
        longPieces.push(longp);
        mixedPieces.push(minp);
        mixedPieces.push(shortp);
        mixedPieces.push(longp);
    }
    // Sorting makes sure if one weekday (or abbr) is a prefix of another it
    // will match the longer piece.
    minPieces.sort(cmpLenRev);
    shortPieces.sort(cmpLenRev);
    longPieces.sort(cmpLenRev);
    mixedPieces.sort(cmpLenRev);
    for (i = 0; i < 7; i++) {
        shortPieces[i] = regex_1.regexEscape(shortPieces[i]);
        longPieces[i] = regex_1.regexEscape(longPieces[i]);
        mixedPieces[i] = regex_1.regexEscape(mixedPieces[i]);
    }
    this._weekdaysRegex = new RegExp('^(' + mixedPieces.join('|') + ')', 'i');
    this._weekdaysShortRegex = this._weekdaysRegex;
    this._weekdaysMinRegex = this._weekdaysRegex;
    this._weekdaysStrictRegex = new RegExp('^(' + longPieces.join('|') + ')', 'i');
    this._weekdaysShortStrictRegex = new RegExp('^(' + shortPieces.join('|') + ')', 'i');
    this._weekdaysMinStrictRegex = new RegExp('^(' + minPieces.join('|') + ')', 'i');
}
