// moment/src/lib/units/week.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const format_1 = require("../format/format");
const aliases_1 = require("./aliases");
const priorities_1 = require("./priorities");
const regex_1 = require("../parse/regex");
const token_1 = require("../parse/token");
const to_int_1 = require("../utils/to-int");
const week_calendar_utils_1 = require("./week-calendar-utils");
// FORMATTING
format_1.addFormatToken('w', ['ww', 2], 'wo', 'week');
format_1.addFormatToken('W', ['WW', 2], 'Wo', 'isoWeek');
// ALIASES
aliases_1.addUnitAlias('week', 'w');
aliases_1.addUnitAlias('isoWeek', 'W');
// PRIORITIES
priorities_1.addUnitPriority('week', 5);
priorities_1.addUnitPriority('isoWeek', 5);
// PARSING
regex_1.addRegexToken('w', regex_1.match1to2);
regex_1.addRegexToken('ww', regex_1.match1to2, regex_1.match2);
regex_1.addRegexToken('W', regex_1.match1to2);
regex_1.addRegexToken('WW', regex_1.match1to2, regex_1.match2);
token_1.addWeekParseToken(['w', 'ww', 'W', 'WW'], function (input, week, config, token) {
    week[token.substr(0, 1)] = to_int_1.default(input);
});
// HELPERS
// LOCALES
function localeWeek(mom) {
    return week_calendar_utils_1.weekOfYear(mom, this._week.dow, this._week.doy).week;
}
exports.localeWeek = localeWeek;
exports.defaultLocaleWeek = {
    dow: 0,
    doy: 6 // The week that contains Jan 1st is the first week of the year.
};
function localeFirstDayOfWeek() {
    return this._week.dow;
}
exports.localeFirstDayOfWeek = localeFirstDayOfWeek;
function localeFirstDayOfYear() {
    return this._week.doy;
}
exports.localeFirstDayOfYear = localeFirstDayOfYear;
// MOMENTS
function getSetWeek(input) {
    var week = this.localeData().week(this);
    return input == null ? week : this.add((input - week) * 7, 'd');
}
exports.getSetWeek = getSetWeek;
function getSetISOWeek(input) {
    var week = week_calendar_utils_1.weekOfYear(this, 1, 4).week;
    return input == null ? week : this.add((input - week) * 7, 'd');
}
exports.getSetISOWeek = getSetISOWeek;
